crates/sim2h/src/demo_im_state.rs

```diff
@@ -0,0 +1,440 @@
+}
+
+impl Default for ConnectionState {
+    fn default() -> Self {
+        Self {
+            con_incr: 0,
+            agent_loc: 0,
+            uri: "".into(),
+        }
+    }
+}
+
+/// so we cache entry locations as well
+#[derive(Debug, Clone)]
+pub struct Entry {
+    entry_loc: u32,
+    aspects: im::HashSet<AspectHash>,
+}
+
+/// sim2h state storage
+#[derive(Debug, Clone)]
+pub struct Store {
+    pub all_aspects: im::HashMap<EntryHash, Entry>,
+    pub connections: im::HashMap<AgentId, ConnectionState>,
+    pub uri_to_connection: im::HashMap<String, AgentId>,
+    pub holding: im::HashMap<AspectHash, im::HashSet<AgentId>>,
+    pub con_incr: Arc<AtomicU64>,
+}
+
+impl Store {
+    pub fn new() -> StoreHandle {
+        let (send_mut, mut recv_mut) = tokio::sync::mpsc::unbounded_channel();
+
+        let con_incr = Arc::new(AtomicU64::new(1));
+        let inner = Arc::new(tokio::sync::Mutex::new(Store {
+            all_aspects: im::HashMap::new(),
+            connections: im::HashMap::new(),
+            uri_to_connection: im::HashMap::new(),
+            holding: im::HashMap::new(),
+            con_incr: con_incr.clone(),
+        }));
+
+        let weak_ref = Arc::downgrade(&inner);
+
+        tokio::task::spawn(async move {
+            let mut should_end_task = false;
+            loop {
+                match weak_ref.upgrade() {
+                    // no more references to us, let this task end
+                    None => return,
+                    Some(strong_ref) => {
+                        // it's fine to await while holding a strong_ref
+                        // just means we may delay a drop of the store
+                        // there is no locking problem
+
+                        match recv_mut.next().await {
+                            // broken channel, let this task end
+                            None => return,
+                            Some(msg) => {
+                                let mut messages = vec![msg];
+
+                                // if we're going to bother locking the mutex
+                                // let's get a few more messages and run
+                                // them all at once
+                                for _ in 0..100 {
+                                    match recv_mut.try_recv() {
+                                        Err(tokio::sync::mpsc::error::TryRecvError::Empty) => break,
+                                        Err(tokio::sync::mpsc::error::TryRecvError::Closed) => {
+                                            should_end_task = true;
+                                            break;
+                                        }
+                                        Ok(msg) => messages.push(msg),
+                                    }
+                                }
+
+                                // now get the lock and perform the mutations
+                                let mut store = strong_ref.lock().await;
```

Comment:
Why do we have to lock the store on get clone? Correct me if I'm wrong but I thought that reads would just use a possibly stale copy.

Reply:
So, I used an Arc<Mutex<>> for ref counting purposes - so the task processing the mutations would just end if there were no references pointing to it.

The only thing that ever acquires the mutex lock is this mutation task, so it should always acquire instantly... I could also use a dummy ref counter and have the state be wholly owned by the mutation task... that's a good idea, I like that.

Reply:
implemented :+1: